zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala

```diff
@@ -466,41 +469,72 @@ private[producer] final class ProducerLive(
-            val length               = serializedRecords.length
-
-            while (it.hasNext) {
-              val (rec, idx): (ByteRecord, Int) = it.next()
-
-              val _ = p.send(
-                rec,
-                (metadata: RecordMetadata, err: Exception) =>
-                  Unsafe.unsafe { implicit u =>
-                    exec {
-                      if (err != null) res(idx) = Left(err)
-                      else res(idx) = Right(metadata)
-
-                      if (count.incrementAndGet == length) {
+        ZIO.suspendSucceed {
+          val recordsLength                                = serializedRecords.length
+          val recordsIterator: Iterator[(ByteRecord, Int)] = serializedRecords.iterator.zipWithIndex
+          val sentResults: Array[Either[Throwable, RecordMetadata]] =
+            new Array[Either[Throwable, RecordMetadata]](recordsLength)
+
+          Ref.make(0).map { sentRecordsCountRef =>
+            @inline def safelyInsertSentResult(resultIndex: Int, sentResult: Either[Throwable, RecordMetadata]): Unit =
+              Unsafe.unsafe { implicit u =>
+                exec {
+                  runtime.unsafe.run(
+                    sentRecordsCountRef.update { sentRecordsCount =>
+                      // Updating sentResults[resultIndex] here is safe:
+                      //  - Ref.update guarantees sentResults.update executed atomically
+                      //  - Ref.update starts with volatile variable read and ends with volatile variable write,
+                      //    which guarantees sentResults.update executed on the latest updated version of sentResults
+                      //    and currently updated version of sentResults
+                      //    will be visible to the next sentResults read or update called within Ref.update
+                      sentResults.update(resultIndex, sentResult)
+
+                      val newSentRecordsCount = sentRecordsCount + 1
+                      if (newSentRecordsCount == recordsLength) {
+                        val sentResultsChunk = Chunk.fromArray(sentResults)
+
                         exec {
-                          runtime.unsafe.run(done.succeed(Chunk.fromArray(res))).getOrThrowFiberFailure()
+                          runtime.unsafe.run(done.succeed(sentResultsChunk))
                         }
                       }
+
+                      newSentRecordsCount
                     }
-                  }
-              )
-            }
-          } catch {
-            case NonFatal(e) =>
-              Unsafe.unsafe { implicit u =>
-                exec {
-                  runtime.unsafe.run(done.succeed(Chunk.fill(serializedRecords.size)(Left(e)))).getOrThrowFiberFailure()
+                  )
+                }
+              }
+
+            var previousSendCallSucceed = true
+
+            while (recordsIterator.hasNext) {
+              val (record: ByteRecord, recordIndex: Int) = recordsIterator.next()
```

Comment:
This is equivalent to:
```suggestion
            recordsIterator.foreach { case (record: ByteRecord, recordIndex: Int) =>
```

Reply:
Good call, applied, 👍